src/swarm-tui/Swarm/TUI/Controller/EventHandlers.hs

```diff
@@ -0,0 +1,131 @@
+import Data.Maybe (fromMaybe, mapMaybe)
+import Data.Text qualified as T
+import Swarm.Game.Failure (SystemFailure (..))
+import Swarm.TUI.Controller.EventHandlers.Frame (runFrameUI, runGameTickUI, ticksPerFrameCap)
+import Swarm.TUI.Controller.EventHandlers.Main (mainEventHandlers)
+import Swarm.TUI.Controller.EventHandlers.REPL (replEventHandlers)
+import Swarm.TUI.Controller.EventHandlers.Robot (handleRobotPanelEvent, robotEventHandlers)
+import Swarm.TUI.Controller.EventHandlers.World (worldEventHandlers)
+import Swarm.TUI.Model
+import Swarm.TUI.Model.Event (SwarmEvent, swarmEvents)
+
+-- ~~~~ Note [how Swarm event handlers work]
+--
+-- Allowing players to customize keybindings requires storing the configuration in AppState.
+-- By doing it as declaratively as possible, Brick also allows us to detect conflicts.
+--
+-- The high level overview is this:
+-- 1. The 'SwarmEvent' is a enumeration of abstracts key events ('PauseEvent', etc.)
+-- 2. The 'AppState' definition contains the key configuration and dispatchers (keys to handlers)
+-- 3. Here in 'Swarm.TUI.Controller.EventHandlers' we declare the handlers for abstract events
+--    and also some non-customizable key handlers (e.g. escape and enter).
+-- 4. When provided with 'KeyConfig' (can include customized keybindings) we can
+--    'createKeyDispatchers' in 'Swarm.TUI.Model.StateUpdate' and store them in 'AppState'.
+-- 5. Finally in 'Swarm.TUI.Controller' the Brick event handler calls the stored dispatchers.
+
+-- | Create key dispatchers that call (abstract) event handlers based on given key config.
+--
+-- Fails if any key events have conflict within one dispatcher or when a main dispatcher
+-- has conflict with one of the subdispatchers.
+createKeyDispatchers ::
+  (Has (Throw SystemFailure) sig m) =>
+  KeyConfig SwarmEvent ->
+  m SwarmKeyDispatchers
+createKeyDispatchers config = do
+  mainGameDispatcher <- buildDispatcher mainEventHandlers
+  let buildSubMainDispatcher = buildSubDispatcher "Main game events" mainGameDispatcher
+  replDispatcher <- buildSubMainDispatcher "REPL panel events" replEventHandlers
+  worldDispatcher <- buildSubMainDispatcher "World view panel events" worldEventHandlers
+  robotDispatcher <- buildSubMainDispatcher "Robot inventory panel events" robotEventHandlers
+  return SwarmKeyDispatchers {..}
+ where
+  -- this error handling code is modified version of the brick demo app:
+  -- https://github.com/jtdaugherty/brick/blob/764e66897/programs/CustomKeybindingDemo.hs#L216
+  buildDispatcher handlers = case keyDispatcher config handlers of
+    Left collisions ->
+      throwLoadingFailure $
+        "Error: some key events have the same keys bound to them.\n"
+          : handlerErrors collisions
+    Right d -> return d
+  buildSubDispatcher parentName parentDispatcher name handlers = do
+    d <- buildDispatcher handlers
+    let collisions = conflicts parentDispatcher d
+    if null collisions
+      then return d
+      else
```

Comment:
May as well have the `conflicts` function return `Maybe (NonEmpty ...)` so that the `NonEmpty`-ness remains attested downstream
```suggestion
    let maybeCollisions = conflicts parentDispatcher d
    case maybeCollisions of
      Nothing -> return d
      Just collisions -> ...
```

Reply:
I support proving with types, but here you run into `brick` not using `NonEmpty` and I want to share code for it.

Unless you suspect `brick` will send us an empty list, I do not see the benefit. The code would have to use `NE.toList` after `NE.nonEmpty`, which seems like a more verbose way to say `null`. 🤔